fix: ensure to call waitUntil with proper "this" if present

maxshirshin opened the pull request:
<!-- Thank you for contributing! -->

### Description

This fixes the way we save a reference to `event.waitUntil` to ensure we preserve the correct `this` when calling

---

### Before submitting the PR, please make sure you do the following:

- [x] Add your change under the `Unreleased` heading in the package's `CHANGELOG.md`
- [x] Read the [Contributing Guidelines](https://github.com/shopify/hydrogen/blob/main/docs/contributing.md)
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`)
- [ ] Update docs in this repository for your change, if needed

jplhomer commented on packages/playground/server-components-worker/worker.js:
```diff
@@ -30,7 +30,7 @@ addEventListener('fetch', (event) => {
         cache: caches.default,
         context: {
-          waitUntil: event.waitUntil,
+            waitUntil: event.waitUntil ? (p) => event.waitUntil(p) : undefined,
```
```suggestion
          waitUntil: event.waitUntil ? (p) => event.waitUntil(p) : undefined,
```

jplhomer reviewed (commented):
Looking good so far - could you update `packages/dev/worker.js`? This is the file that will be used in the starter templates for new Hydrogen apps moving forward.

jplhomer commented:
@maxshirshin oooh wow. Great catch. In that case, could we do that instead in your PR? The file you've touched is for E2E tests only.

maxshirshin commented:
@jplhomer I believe the file I fixed in this PR is the only one which needs to be fixed. `packages/dev/worker.js` looks fine. If the file I changed is for e2e tests only, it still needs to be fixed to avoid being broken after Oxygen ships the `waitUntil` support.
